AsyncDisplayKit/Layout/ASLayoutSpec.mm

```diff
@@ -204,6 +217,11 @@ - (ASEnvironmentTraitCollection)environmentTraitCollection
   return _environmentState.traitCollection;
 }
 
+- (void)setEnvironmentTraitCollection:(ASEnvironmentTraitCollection)traitCollection
+{
+  _environmentState.traitCollection = traitCollection;
```

Comment:
Setting self.environmentState does not cause propagation, so I didn't think this should. This method actually gets called when propagating, so propagating here too would lead to crazy double nested propagation.

Originally there was only the struct based traitCollection. @maicki suggested that we have a nice NSObject based class so that the API was friendly to the end user. That's how ASTraitCollection was born. 

ASEnvironmentTraitCollection sticks around because we are using the ASEnvironment propagation which requires a struct. While I personally like dealing with the struct directly, I think Michael's thought is a noble one: To make these new API more accessible so that people that don't want to deal with a struct don't have to.